### PR TITLE
Interactive Shell Autocomplete Feature

### PR DESCRIPTION
**Pull request by bufgix**
## Description
Added Interactive shell autocomplete feature. This feature provides  autocomplete to custom namespaces or modules 

`help` + `TAB` -> show available commands (already working)
`switch` + `TAB` -> show namespaces
`use`+ `TAB` -> show modules

## How Has This Been Tested?
I  tested on Windows 10.

## Screenshots (if appropriate):
image

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by orhun** [comment]
[Interactive Shell Autocomplete Feature](https://trello.com/c/OLMyUJN3/32-interactive-shell-autocomplete-feature)

**Review by orhun — CHANGES_REQUESTED** [review]
We don't need to import `get_modules` function and call it inside every function that is responsible for auto complete feature since we can access the static modules dictionary with `self.module`. (Which was declared at [here](https://github.com/grapheneX/grapheneX/blob/dev/core/cli/shell.py#L11))
So we should replace the `get_modules` calls with `self.modules` I think.

**Comment by bufgix** [comment]
> We don't need to import `get_modules` function and call it inside every function that is responsible for auto complete feature since we can access the static modules dictionary with `self.module`. (Which was declared at [here](https://github.com/grapheneX/grapheneX/blob/dev/core/cli/shell.py#L11))
> So we should replace the `get_modules` calls with `self.modules` I think.

Fixed https://github.com/grapheneX/grapheneX/pull/25/commits/69625ec960f1a537e25262f8afed7dca99d02d2d

**Review by orhun — APPROVED** [review]

**Review by furk4nyuksel — APPROVED** [review]
tested
 I didn't come across any problems

**Comment by EnesOkutan** [comment]
"use ke" and tab press, not complete. doesn't it have to complete the "kernel" ?
And this completion occurs in a case sensitive way. Is this a problem?

**Review by kalaycitayfun — APPROVED** [review]

File: core/cli/commands.py
Comment (orhun) [inline core/cli/commands.py]: It is not finding the right module with this.
Should be: `AVAILABLE_MODULES.append(f"{key}/{name}")`

Comment (orhun) [reply core/cli/commands.py]: And also I changed the function a little bit for the comfort of my fingers.
Maybe you will consider:
```
    def complete_use(self, text, line, begidx, endidx):
        AVAILABLE_MODULES = self.modules.get(self.namespace)
        if AVAILABLE_MODULES is None:  # if self.namespace == ''
            AVAILABLE_MODULES = list()
            for key, value in self.modules.items():
                for name, module in value.items():
                    AVAILABLE_MODULES.append(f"{key}/{name}")
        mline = line.partition(' ')[2]
        #######
        # If namespace selected
        if '/' in mline:
            # Capitalize given module string for getting rid of case sensitivity
            mline = mline.split('/')[0] + "/" + mline.split('/')[1].capitalize()
        offs = len(mline) - len(text)
        # Get completed text with namespace
        comp_text =  [s[offs:] for s in AVAILABLE_MODULES if s.startswith(mline)]
        # If no namespace found
        if len(comp_text) == 0:
            # Try to complete with module names
            AVAILABLE_MODULES = self.modules.get(list(self.modules.keys())[0])
            mline = mline.capitalize()
            comp_text = [s[offs:] for s in AVAILABLE_MODULES if s.startswith(mline)]
        # Return
        #print(comp_text)
        return comp_text
        #######
```

Comment (orhun) [reply core/cli/commands.py]: Another _nacizane_ idea of mine: Maybe we can move the part that responsible for creating the `AVAILABLE_MODULES` dict. It's just creating that dictionary over and over again on TAB key presses.

Comment (orhun) [reply core/cli/commands.py]: + I've updated the code above.
(Tried to solve case sensitivity issue.)

**Review by EnesOkutan — CHANGES_REQUESTED** [review]
"use Log" + tab does not work.
In the last commit, the entries are converted to lowercase, but the module names in the variable AVAILABLE_MODULES are in the original name.

**Review by afk — APPROVED** [review]

**Comment by orhun** [comment]
> "use Log" + tab does not work.
> In the last commit, the entries are converted to lowercase, but the module names in the variable AVAILABLE_MODULES are in the original name.

Can you test my code above? I've solved that issue.

**Comment by EnesOkutan** [comment]
What code are you talking about?

**Comment by orhun** [comment]
> What code are you talking about?

https://github.com/grapheneX/grapheneX/pull/25#discussion_r279170497

**Comment by EnesOkutan** [comment]
I'm not sure, I don't think it meets exactly.
for example:

use kernel/Logs_R + tab , does not work. 

Since you are applying the capitalize function, the following situation occurs: "kernel/Logs_r"
So the searched module cannot be found in the array.

**Comment by orhun** [comment]
> I'm not sure, I don't think it meets exactly.
> for example:
> 
> use kernel/Logs_R + tab , does not work.
> 
> Since you are applying the capitalize function, the following situation occurs: "kernel/Logs_r"
> So the searched module cannot be found in the array.

Aight, how about this?

`mline = mline.split('/')[0] + "/" + mline.split('/')[1].title()`

**Comment by EnesOkutan** [comment]
Looks fine !

**Comment by orhun** [comment]
```
def complete_use(self, text, line, begidx, endidx):
        avb_modules = self.modules.get(self.namespace)
        if avb_modules is None:
            avb_modules = list()
            for key, value in self.modules.items():
                for name, module in value.items():
                    avb_modules.append(f"{key}/{name}")
        mline = line.partition(' ')[2]
        #######
        # If namespace selected
        if '/' in mline:
            # title() given module string for getting rid of case sensitivity
            mline = mline.split('/')[0].lower() + "/" + mline.split('/')[1].title()
        offs = len(mline) - len(text)
        # Get completed text with namespace
        comp_text =  [s[offs:] for s in avb_modules if s.startswith(mline)]
        # If no namespace found
        if len(comp_text) == 0:
            # Try to complete with module names
            avb_modules = self.modules.get(list(self.modules.keys())[0])
            mline = mline.title()
            comp_text = [s[offs:] for s in avb_modules if s.startswith(mline)]
        # Return
        #print(comp_text)
        return comp_text
        #######
```

**Comment by orhun** [comment]
```
    def complete_switch(self, text, line, begidx, endidx):
        avb_namespaces = [i.lower() for i in self.modules.keys()]
        mline = line.lower().partition(' ')[2]
        offs = len(mline) - len(text)
        return [s[offs:] for s in avb_namespaces if s.startswith(mline)]
```

**Comment by orhun** [comment]
@bufgix Please consider https://github.com/grapheneX/grapheneX/pull/25#issuecomment-487372893 and https://github.com/grapheneX/grapheneX/pull/25#issuecomment-487372966

**Comment by EnesOkutan** [comment]
mline = mline.split('/')[0].lower() + "/" + mline.split('/')[1].title()
In this way, the namespace is case insensitive.

**Comment by orhun** [comment]
> mline = mline.split('/')[0].lower() + "/" + mline.split('/')[1].title()
> In this way, the namespace is case insensitive.

:+1: Edited the code above

**Review by orhun — APPROVED** [review]

**Comment by orhun** [comment]
@EnesOkutan waiting for your approval for merging.

**Comment by EnesOkutan** [comment]
`avb_modules = self.modules.get(list(self.modules.keys())[0])`

I'm not sure this is correct

**Comment by EnesOkutan** [comment]
Doesn't cause error when there is more than one namespace?

**Comment by orhun** [comment]
> `avb_modules = self.modules.get(list(self.modules.keys())[0])`
> 
> I'm not sure this is correct

Why? I don't understand.

**Comment by orhun** [comment]
> Doesn't cause error when there is more than one namespace?

What do you mean "more than one namespace"?

**Comment by orhun** [comment]
I'd be appreciated if you send us the "correct" code. (Personally I don't want to handle more issues on this PR)

**Comment by EnesOkutan** [comment]
This code always gets the first namespace kernel in self.modules. For example, if we think that there is a namespace called "networking", I always think that getting a first module would be an error.

Or is there a misunderstanding, other namespace from kernel namespace

**Comment by orhun** [comment]
> This code always gets the first namespace kernel in self.modules. For example, if we think that there is a namespace called "networking", I always think that getting a first module would be an error.
> 
> Or is there a misunderstanding, other namespace from kernel namespace

Yup, that looks like so. I didn't test the code but it sounds like an error. 
We should get  the `avb_modules` list via `self.namespace` variable.
@bufgix or @EnesOkutan please fix it soon. (Not that hard though)

**Comment by orhun** [comment]
> > This code always gets the first namespace kernel in self.modules. For example, if we think that there is a namespace called "networking", I always think that getting a first module would be an error.
> > Or is there a misunderstanding, other namespace from kernel namespace
> 
> Yup, that looks like so. I didn't test the code but it sounds like an error.
> We should get the `avb_modules` list via `self.namespace` variable.
> @bufgix or @EnesOkutan please fix it soon. (Not that hard though)

nvm -_-

**Comment by EnesOkutan** [comment]
image

I have such a solution but I'm not sure it's the optimal code.

**Review by EnesOkutan — APPROVED** [review]
congratulations, fixed problem.

**Pull request merged**
